### PR TITLE
Patch macdeployqt

### LINKED ISSUE #5487: [package] qt/5.15.2: Plugins not deployed on MacOS
I’m trying to use the conan version of Qt with the following `conanfile.txt`:

```ini
[requires]
qt/5.15.2

[options]
qt:shared=True
qt:qttools=True

[generators]
qt
virtualrunenv
```

When create a bundle application it crash on startup. I realized it was due to the missing plugins as you can see on this terminal output:

```bash
$ export QT_DEBUG_PLUGINS=1
$ ./test.app/Contents/MacOS/test
QFactoryLoader::QFactoryLoader() checking directory path "/Users/martin/Downloads/conan.app/Contents/MacOS/platforms" ...
qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

The problem seems to come from the plugins folder location in the package subfolder (bin/archdatadir ).
The conan recipe offer the possibility to generate a qt.conf file containing information that should help my build chain to work properly:
```ini
[Paths]
Prefix = /Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200
ArchData = bin/archdatadir
HostData = bin/archdatadir
Data = bin/datadir
Sysconf = bin/sysconfdir
LibraryExecutables = bin/archdatadir/bin
Plugins = bin/archdatadir/plugins
Imports = bin/archdatadir/imports
Qml2Imports = bin/archdatadir/qml
Translations = bin/datadir/translations
Documentation = bin/datadir/doc
Examples = bin/datadir/examples 
```
For example, if I copy this file in the `test.app/Contents/Resources` folder, it works.
I also manage to make it work by manually moving the plugins folder from the `bin/archdatadir` package subfolder to the package folder (/Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200).
What is the clean way to handle this issue ?

### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **qt/5.15.2**
  * Operating System+version: **MacOS 10.15.7**
  * Compiler+version: **apple-clang 12.0**
  * Docker image: **N/A**
  * Conan version: **conan 1.36.0**
  * Python version: **Python 3.9.5**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
Configuration for profile default:

[settings]
os=Macos
os_build=Macos
arch=x86_64
arch_build=x86_64
compiler=apple-clang
compiler.version=12.0
compiler.libcxx=libc++
build_type=Release
[options]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)

1. `mkdir build && cd build`
2. `conan install ..`
3. `source activate_run.sh`
4. `qmake ../test.pro`
5. `make`
6. `macdeployqt test.app -verbose=2`

### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```bash
$ conan install ..
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=12.0
os=Macos
os_build=Macos
[options]
[build_requires]
[env]

conanfile.txt: Installing package
Requirements
    brotli/1.0.9 from 'conan-center' - Cache
    bzip2/1.0.8 from 'conan-center' - Cache
    double-conversion/3.1.5 from 'conan-center' - Cache
    freetype/2.10.4 from 'conan-center' - Cache
    libiconv/1.16 from 'conan-center' - Cache
    libjpeg/9d from 'conan-center' - Cache
    libmysqlclient/8.0.17 from 'conan-center' - Cache
    libpng/1.6.37 from 'conan-center' - Cache
    libpq/12.2 from 'conan-center' - Cache
    odbc/2.3.7 from 'conan-center' - Cache
    opengl/system from 'conan-center' - Cache
    openssl/1.1.1j from 'conan-center' - Cache
    pcre2/10.35 from 'conan-center' - Cache
    qt/5.15.2 from 'conan-center' - Cache
    sqlite3/3.33.0 from 'conan-center' - Cache
    zlib/1.2.11 from 'conan-center' - Cache
    zstd/1.4.8 from 'conan-center' - Cache
Packages
    brotli/1.0.9:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Cache
    bzip2/1.0.8:75fe66fbfe61230e9fcbb361f13b6b3baa483e0a - Cache
    double-conversion/3.1.5:6a83d7f783e7ee89a83cf2fe72b5f5f67538e2a6 - Cache
    freetype/2.10.4:f4924bd3b94e155cbf790d74ba3a3c29ccc81e89 - Cache
    libiconv/1.16:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Cache
    libjpeg/9d:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Cache
    libmysqlclient/8.0.17:12e80b538837eca6e02ef1a3eed67aa25cf3423c - Cache
    libpng/1.6.37:88955cec2844f731470e07bd44ce5a3a24ec88b7 - Cache
    libpq/12.2:6615b55b95ff43a02a83c3955d071acd71dad079 - Cache
    odbc/2.3.7:5f4c4ed8ddf27bb70b22f93ac66b6827fc5411d1 - Cache
    opengl/system:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    openssl/1.1.1j:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Cache
    pcre2/10.35:2d8245d819a6b7543095d62df5d99244060071ba - Cache
    qt/5.15.2:0af6aa42f3ece12447af62252e18e9680ffa1200 - Cache
    sqlite3/3.33.0:8ff459f4b248c0e0564ec4e80e52b8d1aca419ce - Cache
    zlib/1.2.11:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Cache
    zstd/1.4.8:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Cache

Installing (downloading, building) binaries...
brotli/1.0.9: Already installed!
bzip2/1.0.8: Already installed!
bzip2/1.0.8: Appending PATH environment variable: /Users/martin/.conan/data/bzip2/1.0.8/_/_/package/75fe66fbfe61230e9fcbb361f13b6b3baa483e0a/bin
double-conversion/3.1.5: Already installed!
libiconv/1.16: Already installed!
libiconv/1.16: Appending PATH environment var: /Users/martin/.conan/data/libiconv/1.16/_/_/package/647afeb69d3b0a2d3d316e80b24d38c714cc6900/bin
libjpeg/9d: Already installed!
opengl/system: Already installed!
openssl/1.1.1j: Already installed!
sqlite3/3.33.0: Already installed!
sqlite3/3.33.0: Appending PATH env var with : /Users/martin/.conan/data/sqlite3/3.33.0/_/_/package/8ff459f4b248c0e0564ec4e80e52b8d1aca419ce/bin
zlib/1.2.11: Already installed!
zstd/1.4.8: Already installed!
libmysqlclient/8.0.17: Already installed!
libpng/1.6.37: Already installed!
libpq/12.2: Already installed!
odbc/2.3.7: Already installed!
odbc/2.3.7: Appending PATH environment variable: /Users/martin/.conan/data/odbc/2.3.7/_/_/package/5f4c4ed8ddf27bb70b22f93ac66b6827fc5411d1/bin
pcre2/10.35: Already installed!
pcre2/10.35: Appending PATH environment variable: /Users/martin/.conan/data/pcre2/10.35/_/_/package/2d8245d819a6b7543095d62df5d99244060071ba/bin
freetype/2.10.4: Already installed!
qt/5.15.2: Already installed!
conanfile.txt: Generator virtualrunenv created activate_run.ps1
conanfile.txt: Generator virtualrunenv created deactivate_run.ps1
conanfile.txt: Generator virtualrunenv created environment_run.ps1.env
conanfile.txt: Generator virtualrunenv created activate_run.sh
conanfile.txt: Generator virtualrunenv created deactivate_run.sh
conanfile.txt: Generator virtualrunenv created environment_run.sh.env
conanfile.txt: Generator qt created qt.conf
conanfile.txt: Generator txt created conanbuildinfo.txt
conanfile.txt: Generated conaninfo.txt
conanfile.txt: Generated graphinfo
$ source activate_run.sh
$ qmake ../test.pro
Info: creating stash file /Users/martin/dev/phonations/Bacasable/build/.qmake.stash
$ make
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -c -pipe -stdlib=libc++ -O2 -std=gnu++11  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.13 -Wall -Wextra -fPIC -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_CORE_LIB -I../../Bacasable -I. -I../../../../.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/include -I../../../../.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/include/QtWidgets -I../../../../.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/include/QtGui -I../../../../.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/include/QtCore -I. -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/System/Library/Frameworks/AGL.framework/Headers -I../../../../.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/bin/archdatadir/mkspecs/macx-clang -o main.o ../main.cpp
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -stdlib=libc++ -headerpad_max_install_names  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.13 -Wl,-rpath,@executable_path/../Frameworks -Wl,-rpath,/Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/lib -o test.app/Contents/MacOS/test main.o   /Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/lib/libQt5Widgets.dylib /Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/lib/libQt5Gui.dylib /Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/lib/libQt5Core.dylib -framework OpenGL -framework AGL
$ macdeployqt test.app -verbose=2
Log:
Log: Deploying Qt frameworks found inside: ("/Users/martin/dev/phonations/Bacasable/build/test.app/Contents/MacOS/test")
Log:  copied: "/Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/lib/libQt5Widgets.5.dylib"
Log:  to "test.app/Contents/Frameworks//libQt5Widgets.5.dylib"
Log:  copied: "/Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/lib/libQt5Gui.5.dylib"
Log:  to "test.app/Contents/Frameworks//libQt5Gui.5.dylib"
Log:  copied: "/Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/lib/libQt5Core.5.dylib"
Log:  to "test.app/Contents/Frameworks//libQt5Core.5.dylib"
Log:
Log: Deploying plugins from "/Users/martin/.conan/data/qt/5.15.2/_/_/package/0af6aa42f3ece12447af62252e18e9680ffa1200/plugins"
Log: Created configuration file: "test.app/Contents/Resources/qt.conf"
Log: This file sets the plugin search path to "test.app/Contents/PlugIns"
```

</details>

### PR DESCRIPTION
**Pull request by MartinDelille**
Specify library name and version:  **qt/5.15.2**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

fix #5487

### CODE REVIEW

File: recipes/qt/5.x.x/patches/fix-macdeployqt.diff [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
++#include <QLibraryInfo>
+ 
+ #include "../shared/shared.h"
+ 
+@@ -236,7 +237,7 @@ int main(int argc, char **argv)
+     }
+ 
+     if (plugins && !deploymentInfo.qtPath.isEmpty()) {
+-        deploymentInfo.pluginPath = deploymentInfo.qtPath + "/plugins";
++        deploymentInfo.pluginPath = QLibraryInfo::location(QLibraryInfo::PluginsPath);
```
Comment (Minimonium) [inline recipes/qt/5.x.x/patches/fix-macdeployqt.diff:18]: From the visible design perspective - I think it'd be better to fill the `pluginPath` from the `shared.cpp`, at the same place as the `qtPath`. 
Then, here you'd need to only check that `!deploymentInfo.pluginPath.isEmpty()`.

**Comment by ericLemanissier** [comment]
Thanks ! Could you please report the bug at bugreports.qt.io, and submit the patch for review at codereview.qt.io ?

**Comment by MartinDelille** [comment]
The issue has already been adressed here: https://github.com/qt/qttools/commit/03abcbabbd4caa11048d19d95b23f165cd7a5361

Shall I generate the patch from this commit instead ?

Do we agree that the CI failure has nothing to do with my change ?

**Comment by Minimonium** [comment]
> Shall I generate the patch from this commit instead ?
That would be ideal.

**Comment by MartinDelille** [comment]
Unfortunately, the patch doesn't apply correctly, probably due to the fact that other commits have modified the file:

```bash
git log --oneline v5.15.2..dev src/macdeployqt/macdeployqt/main.cpp
03abcbabb macdeployqt: Fix plugin resolution bugs for non-standard installs
e6ad15080 Merge remote-tracking branch 'origin/5.15' into dev
b2f3eb655 Merge remote-tracking branch 'origin/5.15' into dev
6b2f32f3d Fix Qt6 build
```

**Comment by MartinDelille** [comment]
@ericLemanissier I fixed the patch because it looks like the *QLibraryInfo* class changed in Qt 6:
- qt5: https://doc.qt.io/qt-5/qlibraryinfo.html#location
- qt6: https://doc.qt.io/qt-6/qlibraryinfo.html#path

Should I submit a patch to the 5.15 branch ?

**Comment by MartinDelille** [comment]
Is there missing for this PR to be accepted ?

**Comment by prince-chrismc** [comment]
My only question is will there be another 5.x.x release that will include this fix in the near future?

If no, I'll approve 🙃

**Comment by wdobbe** [comment]
@prince-chrismc Qt company has announced that updates to qt 5.15 will only available to commercial customers. So it is not very likely that we will get a new 5.x release. Although I think KDE maintains a Qt 5.15 fork with patches.

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by ericLemanissier — APPROVED** [review]

**Pull request merged**